drv/update-api/src/lib.rs

```diff
@@ -87,6 +89,10 @@ impl From<UpdateError> for GwUpdateError {
             UpdateError::ImageBoardUnknown => Self::ImageBoardUnknown,
             UpdateError::TaskRestarted => Self::TaskRestarted,
             UpdateError::NotImplemented => Self::NotImplemented,
+
+            // TODO: Update gateway error to handle this then change the error to
+            // a proper enum
+            UpdateError::MissingHandoffData => Self::Unknown(value as u32),
```

Comment:
I'm starting to realize this strategy is not going to work. We won't actually know when all MGSs are updated with the latest errors and so if we send back something other than Unknown(error_code), MGS may not be able to deserialize it.

I think what we actually want to do here is to update the error in `gateway-messages` first, and use it correctly as a variant in this code. Then in MGS itself, when an error is returned, we check to see if the discriminant is known. If it's not we fill in the discriminant value received on the wire. This makes deserialization messier than we like, but preserves error codes.

Thoughts here @mkeeter @cbiffle @jgallagher ?

Reply:
My vote would be to update the error in `gateway-messages` (bumping the version there), then use the new error type here.

I'm ambivalent about whether MGS should jump through hoops to convert a failed-to-deserialize `UpdateError` into `gateway_messages::UpdateError::Unknown`, versus just failing deserialization; the latter seems more correct to me, and would produce an obvious error ("deserialization failed, and the SP MGS version is newer than ours").

Reply:
Matt's reply seems right to me, but I'm not entirely sure what you meant by

> Then in MGS itself, when an error is returned, we check to see if the discriminant is known. If it's not we fill in the discriminant value received on the wire. 

What does "check to see if the discriminant is known" look like given we're hubpack-deserializing the entire message? It will fail on an unknown variant. I guess we could do hubpack-within-hubpack? The outer message is something like `UpdateError(u32)`, and then we attempt to hubpack-deserialize the `u32` into a more specific `UpdateErrorCode` as a second step?

Reply:
Yeah, it's just an error, and a nearly impossible one at that.  I'll do as Matt suggests.

> Matt's reply seems right to me, but I'm not entirely sure what you meant by

>> Then in MGS itself, when an error is returned, we check to see if the discriminant is known. If it's not we fill in the discriminant value received on the wire.

What I meant is that we could actually write code to attempt to deserialize the error, and if not parse it with a different hubpack message that short circuits all errors as `Unknown(u32)`. I don't actually think that would work though because the discriminants in hubpack are only 8 bits, and so hubpack deserialization would fail. Anyway, this all sounds insane, so skip it.

You're idea about inner/outer isn't a bad one, but also likely overkill.

Reply:
Fixed in [3c3ca2e](https://github.com/oxidecomputer/hubris/pull/1367/commits/3c3ca2ea4042426ff64a3f7f6f947fb5f910f10c)